python/paddle/fluid/contrib/mixed_precision/fp16_utils.py

```diff
@@ -115,6 +115,66 @@ def _insert_cast_op(block, op, idx, src_dtype, dest_dtype):
     return num_cast_ops
 
 
+def check_next_input_backward_role_op(ops, idx, cur_op, var_name):
+    """
+    Check whether has op (backward_role) that inputs var_name 
+    variable behind cur_op.
+    If yes, it means _move_optimize_ops_back will cause errors
+    in program order. Therefore raise valueerror.
+    If no, return True to continue move in _move_optimize_ops_back.
+
+    Args:
+        ops (list): A list of ops.
+        id (int): index of cur_op in ops
+        cur_op (Operator): Current operator which has var_name variable.
+        var_name (string): Variable name.
+    """
+    BACKWARD = core.op_proto_and_checker_maker.OpRole.Backward
+    next_op = []
+    for i in range(idx + 1, len(ops)):
+        op = ops[i]
+        if op.attr('op_role') & int(BACKWARD):
+            for input_name in op.input_arg_names:
+                if input_name == var_name:
+                    next_op.append(op)
+    if next_op:
+        raise ValueError("There must be no next op that inputs {0} "
+                         "variable after {1} op (optimize_role)".format(
+                             var_name, cur_op.type))
+    return True
+
+
+def _move_optimize_ops_back(block):
+    """
+    put optimize_role ops(cast) behind the backward_role ops
+    to speed up in Executor.
+
+    Args:
+        block: The global_block of main program. eg: block = main_prog.global_block()
+    """
+    OPTIMIZE = core.op_proto_and_checker_maker.OpRole.Optimize
+    optimize_ops = []
+    for idx, op in reversed(list(enumerate(block.ops))):
+        if op.attr('op_role') & int(OPTIMIZE):
+            for output_name in op.output_arg_names:
+                if check_next_input_backward_role_op(block.ops, idx, op,
```

Comment:
why not check all output vars in one call?

Reply:
Done. Thanks!